---
title: "ChIP-seq analysis"
author: "Vanessa Dumeaux"
output: rmarkdown::html_notebook
---

#  1. Unsupervised analysis - MACS2 peaks
We conducted an unsupervised analysis of peaks computed individually for each sample by [MACS2](https://github.com/taoliu/MACS) using the DiffBind R package 

Refs: 

+ [Stark R and Brown G (2011). DiffBind: differential binding analysis of ChIP-Seq peak data.](http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf)

+ [Ross-Innes CS, Stark R, Teschendorff AE, Holmes KA, Ali HR, Dunning MJ, Brown GD, Gojis O, Ellis IO, Green AR, Ali S, Chin S, Palmieri C, Caldas C, Carroll JS (2012). “Differential oestrogen receptor binding is associated with clinical outcome in breast cancer.” Nature, 481,-4.](http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html)

We precomputed the DBA object to read in peaksets output from MACS2. Please refer to the Diffbind R package to format `sample.diffbind` that provides paths to the peak calls from MACS2.

```{r, message=FALSE}
library(GenomicRanges)
library(DiffBind)
```
```{r, eval=FALSE}
t30.dba <- dba(sampleSheet = sample.table[sample.diffbind$Treatment=="t30" & !sample.diffbind$Factor=="SUMO",])
t180.dba <- dba(sampleSheet = sample.table[sample.diffbind$Treatment=="t180" & !sample.diffbind$Factor=="SUMO",])

t30.dba.counts <- dba.count(t30.dba)
t180.dba.counts <- dba.count(t180.dba)
```
##### + Exploratory plots of ChIP-seq peaks

###### PCA plot of log2-normalized counts
```{r, eval=FALSE}
```
```{r}
t30.dba.counts <- readRDS("../data/t30.dba.counts.rds")
dba.plotPCA(t30.dba.counts,attributes = DBA_CONDITION, dotSize = 0.8)

t180.dba.counts <- readRDS("../data/t180.dba.counts.rds")
dba.plotPCA(t180.dba.counts,attributes = DBA_CONDITION, dotSize = 0.8)
```

##### + Exploratory plots of ChIP-seq SUMO

###### PCA plot of log2-normalized counts
```{r, eval=FALSE}
sumo.dba <- dba(sampleSheet=sample.diffbind[sample.diffbind$Factor=="SUMO",])
sumo.dba.counts <- dba.count(sumo.dba)
```
```{r}
sumo.dba.counts <- readRDS("../data/sumo.dba.counts.rds")
dba.plotPCA(sumo.dba.counts,attributes = c(DBA_CONDITION,DBA_TREATMENT), dotSize = 0.8)
```

# 2. ERE analysis

##### + Import ERE regions coordinates
```{r, message=FALSE}
library(GenomicRanges)

ere <- read.table("../data/EREs.hg19.txt", sep = "\t", 
                  header = T, stringsAsFactors = FALSE)
colnames(ere) <- ifelse(colnames(ere)=="Chrom.", "chrom", colnames(ere))
ere_hg19 <- makeGRangesFromDataFrame(ere)
```

##### + Identify ERa peaks (overlap between E2-30' and ICI-30') that contain ERE
```{r}
peaks <- read.table("../data/peaks-overlap.txt", sep="\t", header = T)
head(peaks)
peaks$idx <- 1:nrow(peaks)
dim(peaks)

E2ICIpeaks <- GenomicRanges::makeGRangesFromDataFrame(peaks,
                                      start.field = "starts",
                                      end.field = "ends")
E2ICIpeaks
rtracklayer::export.bed(E2ICIpeaks, "../data/E2ICIpeaks.bed")

olap <- findOverlaps(ere_hg19, E2ICIpeaks, type="within")
olap

rtracklayer::export.bed(E2ICIpeaks[olap@to,], "../data/E2ICIpeaks-ERE.bed")
```
##### + Plot tag density heatmap using deepTools2
Trimmed and sorted bam files were normalized using [deepTools2 version 2.0](https://deeptools.readthedocs.io/en/2.0.0/content/list_of_tools.html)
```{bash, eval = FALSE}
#!/usr/bin/env bash
# File: batch-bamCoverage.sh
files=`find . -name "*.sorted.dup.bam"`

for i in $files
do
out=`basename "${i%%.sorted.dup.bam}"`
echo $out
bamCoverage -b $i -o normalizeTo1/$out.norm.bw \
--normalizeTo1x 2451960000 --maxFragmentLength 600 -p 40
done
```
`computeMatrix()` calculates scores per regions of interest (ERa peaks containing EREs)  +/- 3kb and `plotHeatmap()` draws the tag density for these regions.

```{bash, eval = FALSE}
computeMatrix reference-point \
    -S  V_30_merged.bw E2_30_merged.bw ICI_30_merged.bw ICI_180_merged.bw \
    -R ~/repositories/mader/chipseq-ICI/data/E2ICIpeaks-ERE.bed \
    -b 3000 -a 3000 \
    -o ../../ERpeaksERE-chipER-refpoint.gz \
    --referencePoint center \
    -p 40
```
```{bash, eval = FALSE}
plotHeatmap -m ERpeaksERE-chipER-refpoint.gz \
    -out heatmap-E2ICIpeaks-ERE-chipER-refpoint.pdf \
    --sortUsingSamples 1 \
    --colorList 'white,blue'
```



# 3. Study overlap between ChIP-seq ERa and SUMO peaks

```{r, message=FALSE}

library(ggplot2)

q.plot <- ggplot(peaks, aes(x = idx, y=qvalue..log10))+
  theme_bw() +
  #ggtitle("")+
  ylab("-log10 q-value")+
  geom_point()

print(q.plot)

### show proportion of ICI-SUMO overlap by quantiles of top peaks
q <- quantile(peaks$qvalue..log10, probs = c(0.95, 0.90, 0.80, 0.50, 0.25, 0))
q

peaks$ICI.30.ERa <- 1

library(limma)
a <- unlist(lapply(q, function(x){
  tab <- vennCounts(peaks[peaks$qvalue..log10 > x, colnames(peaks) %in% c("V.30.SUMO","ICI.30.ERa")])
  prop <- tab[4,3]/(tab[2,3]+tab[4,3])
  return(prop)
  }))

b <- unlist(lapply(q, function(x){
  tab <- vennCounts(peaks[peaks$qvalue..log10 > x,colnames(peaks) %in% c("ICI.30.SUMO","ICI.30.ERa")])
  prop <- tab[4,3]/(tab[2,3]+tab[4,3])
  return(prop)
  }))

prop <- c(a, b)
quant.labels <- c("top 5%", "top 10%", "top 20%", "top 50%", "top 75%", "all")
q.labels <- factor(rep(quant.labels, 2), levels=quant.labels, ordered = TRUE)
quant <- c(names(a), names(b))
exp <- c(rep("V.30.SUMO", length(a)),
         rep("ICI.30.SUMO", length(b)))
mydata <- data.frame(value = prop,
                     quant = quant,
                     q.labels = q.labels,
                     exp = exp)

p <-ggplot(mydata, aes(q.labels, value))+
  geom_bar(stat = "identity", aes(fill = exp), position = "dodge") +
  xlab("top ERa-ICI-30 min peaks (ordered by -log10 q-value)") + ylab("Proportion of overlap") +
  ggtitle("Overlap with ERa ICI 30min peaks") +
  theme_bw()

p
```

##### + Select top 5% ERa peaks and draw normalized tag density heatmaps in SUMO experiments
```{r, message=FALSE}
top.ERa.peaks <- peaks[peaks$qvalue..log10 > q[1], ]

top.ERa.peaks <- GenomicRanges::makeGRangesFromDataFrame(top.ERa.peaks,
                                      start.field = "starts",
                                      end.field = "ends")
top.ERa.peaks

rtracklayer::export.bed(top.ERa.peaks, "../data/top-ERa-peaks.bed")
```

##### + Plot tag density heatmap using deepTools2
```{bash, eval = FALSE}
computeMatrix reference-point \
    -S V_30_SUMO_merged.bw ICI_30_SUMO_merged.bw ICI_180_SUMO_merged.bw \
    -R ~/repositories/mader/chipseq-ICI/data/sel.coord.bed \
    -b 3000 -a 3000 \
    -o ../../topERpeaks-chipSUMO-refpoint.gz \
    --referencePoint center \
    -p 40
```
```{bash, eval = FALSE}
plotHeatmap -m topERpeaks-chipSUMO-refpoint.gz \
    -out heatmap-topERpeaks-chipSUMO-refpoint.pdf \
    --sortUsingSamples 1 \
    --colorList 'white,blue'
```

# 4. Reproducibility of SUMO binding for topER peaks

```{r}
corr2 <- read.table("~/Downloads/topERpeaks-allSUMO-compare.tab", sep="\t")
colnames(corr2) <- c("seqnames", "start", "end", "V_30_SUMO_N1", "V_30_SUMO_N2",
                    "ICI_30_SUMO_N1", "ICI_30_SUMO_N2", "ICI_3h_SUMO_N1", "ICI_3h_SUMO_N2")
head(corr2)


p <- ggplot(corr2, aes(x=ICI_30_SUMO_N1, y=ICI_30_SUMO_N2)) +
  geom_point() +
  theme_bw()
p
```{r, echo=FALSE}
pdf("~/Dropbox/mader/Oncogene Traboulsi et al/3rd RESUBMISSION/3-FigureS6D-reproduciility/scatterplot-topERpeaks-SUMOICI30.pdf")
print(p)
dev.off()
```
```{r}
cor.test(corr2$ICI_30_SUMO_N1, corr2$ICI_30_SUMO_N2, method="spearman")
cor.test(corr2$ICI_30_SUMO_N1, corr2$ICI_30_SUMO_N2)
```

```{r}

p <- ggplot(corr2, aes(x=V_30_SUMO_N1, y=V_30_SUMO_N2)) +
  geom_point() +
  theme_bw()
p

pdf("~/Dropbox/mader/Oncogene Traboulsi et al/3rd RESUBMISSION/3-FigureS6D-reproduciility/scatterplot-topERpeaks-SUMOV30.pdf")
print(p)
dev.off()

cor.test(corr2$V_30_SUMO_N1, corr2$V_30_SUMO_N2, method="spearman")

cor.test(corr2$V_30_SUMO_N1, corr2$V_30_SUMO_N2)
```
```{r}

p <- ggplot(corr2, aes(x=ICI_3h_SUMO_N1, y=ICI_3h_SUMO_N2)) +
  geom_point() +
  theme_bw()
p

pdf("~/Dropbox/mader/Oncogene Traboulsi et al/3rd RESUBMISSION/3-FigureS6D-reproduciility/scatterplot-topERpeaks-SUMOICI3h.pdf")
print(p)
dev.off()

cor.test(corr2$ICI_3h_SUMO_N1, corr2$ICI_3h_SUMO_N2, method="spearman")
cor.test(corr2$ICI_3h_SUMO_N1, corr2$ICI_3h_SUMO_N2)
```





